pallets/collator-selection/src/lib.rs

```diff
@@ -499,29 +512,83 @@ pub mod pallet {
+				.saturating_sub(1); // Note: -1 to accomodate 0-index counting // RAD: Is there API to saturate but notify on non-overflow so we can log a warning?
+					// 3. Block number at rank is the percentile and our kick performance benchmark
+			let blocks_created_at_percentile: BlockCount =
+				collator_perf_this_session[ordinal_rank].1; // XXX: don't like the tuple accessor, could this be a struct?
+											// 4. We kick if a collator produced UnderperformPercentileByPercentToKick fewer blocks than the percentile
+			let threshold_factor =
+				1.0 - T::UnderperformPercentileByPercentToKick::get() as f64 / 100.0;
+			let kick_threshold =
+				(threshold_factor * (blocks_created_at_percentile as f64)) as BlockCount;
+			log::info!(
+				"Session Performance stats: {}-th percentile: {} blocks\nWill kick under {} blocks",
+				T::PerformancePercentileToConsiderForKick::get(),
+				blocks_created_at_percentile,
+				kick_threshold
+			);
+
+			// 5. Walk the percentile slice, call try_remove_candidate if a collator is under threshold
+			let current_candidate_ids = candidates
 				.into_iter()
-				.filter_map(|c| {
-					let last_block = <LastAuthoredBlock<T>>::get(c.who.clone());
-					let since_last = now.saturating_sub(last_block);
-					if since_last < kick_threshold {
-						Some(c.who)
-					} else {
-						let outcome = Self::try_remove_candidate(&c.who);
-						if let Err(why) = outcome {
-							log::warn!("Failed to remove candidate {:?}", why);
-							debug_assert!(false, "failed to remove candidate {:?}", why);
-						}
-						None
+				.map(|i: CandidateInfo<T::AccountId, BalanceOf<T>>| i.who.clone())
+				.collect::<Vec<_>>();
+			let mut removed_account_ids: Vec<T::AccountId> = Vec::new();
+			let kick_candidates = collator_perf_this_session[..ordinal_rank] // ordinal-rank exclusive, the collator with percentile perf is safe
+				.iter()
+				.map(|acc_info| acc_info.0.clone())
+				.collect::<Vec<_>>();
+			kick_candidates.into_iter().for_each(|acc_id| {
+				// If we're not a candidate we're invulnerable or already kicked
+				if current_candidate_ids.contains(&acc_id) {
+					let my_blocks_this_session = <BlocksPerCollatorThisSession<T>>::get(&acc_id); // RAD: read storage or find by iterating collator_perf_this_session vec, which is faster
+					if my_blocks_this_session <= kick_threshold {
+						Self::try_remove_candidate(&acc_id)
+							.and_then(|_| {
+								removed_account_ids.push(acc_id.clone());
+								Ok(())
+							})
+							.unwrap_or_else(|why| -> () {
+								log::warn!("Failed to remove candidate {:?}", why);
+								debug_assert!(false, "failed to remove candidate {:?}", why);
+							});
 					}
-				})
-				.collect::<Vec<_>>()
+				}
+			});
+			removed_account_ids
+		}
+
+		/// Reset the performance map to the currently active validators at 0 blocks
+		pub fn reset_collator_performance() {
+			// XXX: intersect BPC map with validators and try_mutate to 0 oder or drop and recreate from scratch?
```

Comment:
Should we intersect the BlocksPerCollatorThisSession map with validators() and try_mutate to 0 or drop the map with remove_all and recreate from scratch?

Latter is much more concise, don't know about performance of the former